docs/infrastructure/sbt.md

```diff
@@ -14,29 +14,88 @@ compilation. The build configuration is defined in
 [zinc](https://github.com/sbt/zinc) (the incremental compiler for Scala). zinc
 analyses the compiled files and detects dependencies between them to determine
 which files have to be recompiled when something has been changed.
 
-## Helper tasks
+## Bootstrapping
+As described in [Java 11 Migration](./java-11.md#illegalaccesserror) to
+successfully compile the `runtime` project, the JVM running sbt must use the
+overridden JAR for Truffle API. This JAR has to be present during startup of the
+JVM, but it has to be downloaded from the Maven repository.
+
+To fix this chicken-and-egg problem, we have a special `bootstrap` task, that
+has to be ran when setting-up the project (and after a version change of Graal).
+It makes sure the JAR is downloaded and copied to our directory and terminates
+the sbt process to ensure that the user restarts it. Without the full restart,
+the JAR would not be seen by the JVM. So when setting up the project or after
+changing the version of Graal, before launching the sbt shell, you should first
+run `sbt bootstrap`, to make sure the environment is properly prepared.
+
+The logic for copying the JAR is implemented in the `bootstrapJARs` task in
+[`CopyTruffleJAR`](../../project/CopyTruffleJAR.scala).
+
+## Compile Hooks
+There are some invariants that are specific to our project, so they are not
+tracked by sbt, but we want to ensure that they hold to avoid cryptic errors at
+compilation or runtime.
+
+To check some state before compilation, we add our tasks as dependencies of
+`Compile / compile / compileInputs` by adding the following to the settings of a
+particular project.
+
+```
+Compile / compile / compileInputs := (Compile / compile / compileInputs)
+        .dependsOn(preCompileHookTask)
+        .value
+```
+
+Tasks that should be run before compilation, should be attached to the
+`compileInputs` task. That is because the actual compilation process is ran in
+the task `compileIncremental`. `Compile / compile` depends on
+`compileIncremental` but if we add our dependency to `Compile / compile`, it is
+considered as independent with `compileIncremental`, so sbt may schedule it to
+run in parallel with the actual compilation process. To guarantee that our
+pre-flight checks complete *before* the actual compilation, we add them as a
+dependency of `compileInputs` which runs *strictly before* actual compilation.
+
+To check some invariants *after* compilation, we can replace the original
+`Compile / compile` task with a custom one which does its post-compile checks
+and returns the result of `(Compile / compile).value`. An example of such a
+'patched' compile task is implemented in
+[`FixInstrumentsGeneration`](../../project/FixInstrumentsGeneration.scala).
+
+## Helper Tasks
 There are additional tasks defined in the [`project`](../../project) directory.
 They are used by [`build.sbt`](../../build.sbt) to provide some additional
 functionality.
 
+### GraalVM version check
```

Comment:
You should use title case for section titles. `GraalVM Verxsion Check`. Please check that all titles conform to this.